### ISSUE TITLE
V11 comtability: Ringct type 3 is not allowed from v11 

### ISSUE THREAD

**Issue by moneroexamples** [mymonero/mymonero-app-js#277]
Demon produces error: 
```
2019-02-28 10:32:53.305 [RPC0]  TRACE   blockchain      src/cryptonote_core/blockchain.cpp:2351 Blockchain::check_tx_outputs                                                                            
2019-02-28 10:32:53.305 [RPC0]  ERROR   verify  src/cryptonote_core/blockchain.cpp:2437 Ringct type 3 is not allowed from v11                                                                           
2019-02-28 10:32:53.307 [RPC0]  INFO    txpool  src/cryptonote_core/tx_pool.cpp:213     Transaction with id= <3906846c2faddeb21a2d9499673c62c579b78547652ec91fcf28f45c1714f1a1> has at least one invalid output
2019-02-28 10:32:53.307 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        2      add_tx                                                                                               
2019-02-28 10:32:53.307 [RPC0]  ERROR   verify  src/cryptonote_core/cryptonote_core.cpp:960     Transaction verification failed: <3906846c2faddeb21a2d9499673c62c579b78547652ec91fcf28f45c1714f1a1>     
2019-02-28 10:32:53.307 [RPC0]  WARN    daemon.rpc      src/rpc/core_rpc_server.cpp:721 [on_send_raw_tx]: tx verification failed: invalid output                                                        
2019-02-28 10:32:53.307 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF       18    on_send_raw_tx                                                                                         
2019-02-28 10:32:53.307 [RPC0]  DEBUG   net.http        src/rpc/core_rpc_server.h:96    /sendrawtransaction processed with 0/18/0ms  
```

when trying to send txs on v11 stagenet.

**Comment by paulshapiro** [comment]
mymonero-app-js, core-js, core-cpp, libapp-js, and libapp-cpp have been updated with 0.14.0.0 compat. Can't test on mainnet yet as tx will fail submission validation (invalid output).

**Comment by moneroexamples** [comment]
Thank you.

I see new PR with the changes merged just recently: https://github.com/mymonero/mymonero-app-js/commit/86039bdd14d7c2b2c879cce2ea464e1814a087e5

I will check on stagenet, as I already migrated all my stagenet nodes to v11, and will report back.

**Comment by moneroexamples** [comment]
Little update. Now it sometimes works, but sometimes does not. 

Its the same error as before, for txs that fail. Don't know yet, why some txs are different than others, why some fail.

**Comment by moneroexamples** [comment]
Its different error after all. Orginal error was about invalid outputs. Some, as in most, fail do invalid input:

```
019-03-01 03:24:08.416 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        0          expand_transaction_2
510893 2019-03-01 03:24:08.422 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        5            verRctMGSimple
510894 2019-03-01 03:24:08.422 [RPC0]  INFO    ringct  src/ringct/rctSigs.cpp:1066 verRctMGSimple failed for input 0
510895 2019-03-01 03:24:08.422 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        5          verRctNonSemanticsSimple
510896 2019-03-01 03:24:08.422 [RPC0]  ERROR   verify  src/cryptonote_core/blockchain.cpp:2836 Failed to check ringct signatures!
510897 2019-03-01 03:24:08.422 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        7        check_tx_inputs
510898 2019-03-01 03:24:08.422 [RPC0]  INFO    txpool  src/cryptonote_core/tx_pool.cpp:266 tx used wrong inputs, rejected
510899 2019-03-01 03:24:08.422 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF        8      add_tx
510900 2019-03-01 03:24:08.422 [RPC0]  ERROR   verify  src/cryptonote_core/cryptonote_core.cpp:960 Transaction verification failed: <5a0845b47c4bd143cad71219df33e606b6a222fdd1c3b99a6e7d62dce7ea62b2>
510901 2019-03-01 03:24:08.422 [RPC0]  WARN    daemon.rpc  src/rpc/core_rpc_server.cpp:721 [on_send_raw_tx]: tx verification failed: invalid input
510902 2019-03-01 03:24:08.422 [RPC0]  DEBUG   perf    src/common/perf_timer.cpp:140   PERF       27    on_send_raw_tx
510903 2019-03-01 03:24:08.422 [RPC0]  DEBUG   net.http    src/rpc/core_rpc_server.h:96    /sendrawtransaction processed with 0/28/0ms
```

**Comment by moneroexamples** [comment]
I prepared example of tx that fails. Maybe will be useful.

The tx in hex is here: https://paste.fedoraproject.org/paste/OikdL7u2fwdEicTvTglbIg

The view this tx in decoded form, please copy and paste the hex into the tx pusher of the onion explorer, e.g.: http://139.162.60.17:8082/rawtx and press Check. Trying to Push the tx, will result in invalid input error.

For comparison, this is example of tx that was accepted: http://139.162.60.17:8082/tx/b4e9b185fc03949b9982f4474831febb8a1827cfae1caa26356c9e75fb6801e5

**Comment by luigi1111** [comment]
@moneroexamples is it the same inputs that fail, or just that they all fail most of the time?

**Comment by moneroexamples** [comment]
@luigi1111 

yes. I have some suspicion now. I see now that 

```
  "ecdhInfo": [ {
        "amount": "c99a65ab35b70809"
      }, {
        "amount": "231f5c0545e1dc1b"
      }], 
```
Are short now. Before they were longer. I also see that ecdhInfo `mask` is gone for example. I think that I incorrectly handle this in openmonero. Subsequently wrong `rct` value is provided to mymonero.

Based on the api docks: https://github.com/monero-project/meta/blob/c139cb39ea56441fe7cb7ff13b74fe34b394f2f5/api/lightwallet_rest.md#import_request

> rct is, for ringct outputs, a 96-byte blob containing the concatenation of the public commitment, then the ringct mask value, and finally the ringct amount value. For ringct coinbase outputs, the mask is always the identity mask and the amount is zero; for non-coinbase ringct outputs, the mask and amount are the respective raw encrypted values, which must be decrypted by the client using the view secret key. For non-ringct outputs, this field is nil.

So I think the question is, has the format changed now? 

And txs that were successful, might have used only coinbase txs.

**Comment by luigi1111** [comment]
Ah, that makes perfect sense.

That's actually a little annoying to handle I guess. Perhaps you could just send the field as either 96 or 40 bytes and the client could interpret based on length.

Otherwise you'd need some kind of versioning, or the API/server could calculate the values and send them the old way (that sounds pretty meh though).

**Comment by moneroexamples** [comment]
Yes, this seems to be the case. 

openmonero is only backend to mymonero frontend, so I need to try to handle this in a way mymonero expects it. For now I don't know if mymonero requires some other or new type/format of information in the `rtc` fields in its api or not?

**Comment by moneroexamples** [comment]
@paulshapiro 

Do you know if there were any changes made to `rct` fields in mymonero backend, and what are these fields now?

**Comment by paulshapiro** [comment]
@moneroexamples we havent changed anything on the server yet. Still digging but haven't diagnosed yet. What we know so far is that `monero_transfer_utils::_rct_hex_to_rct_commit` *could* be updated for 40 byte fields… but it's actually functionally OK as it is because it's looking at the head of the string… and it also wouldn't cause an issue of 'invalid input' afaik. Does seem that `sc_isnonzero` from `MLSAG_Ver` must be returning false. We're continuing work to debug.

**Comment by moneroexamples** [comment]
@paulshapiro 

Thank you for update. I though that its already working for you, and it only does not work for openmonero. I also keep investigating on what's happening.

**Comment by moneroexamples** [comment]
Hi
Got it working. Still need to fully test it, but now I dont see the issue anymore.

The solution (hopefully) seems to be to simply provide decrypted rtc_mask mask to the fronted. Subsequently, the frontend don't need to decrypt it anymore.



Patch that I used in `mymonero-core-js`:

```
diff --git a/src/monero_transfer_utils.cpp b/src/monero_transfer_utils.cpp
index 4d11a06..0428d47 100644
--- a/src/monero_transfer_utils.cpp
+++ b/src/monero_transfer_utils.cpp
@@ -154,6 +154,11 @@ bool _rct_hex_to_decrypted_mask(
        std::string encrypted_mask_str = rct_string.substr(64,64);
        THROW_WALLET_EXCEPTION_IF(!string_tools::validate_hex(64, encrypted_mask_str), error::wallet_internal_error, "Invalid rct mask: " + encrypted_mask_str);
        string_tools::hex_to_pod(encrypted_mask_str, encrypted_mask);
+
+    std::cout << "dont decrypt mask as now the decrypted mask is provided by backend" << endl;
+    decrypted_mask = encrypted_mask;
+    return true;
+
        //
        if (encrypted_mask == rct::identity()) { // NOTE: ringct coinbase txs have the identity mask manually provided unencrypted in the rct field by the hosted lightwallet backend
                decrypted_mask = encrypted_mask;
```

And on the backend I simply use decrypted rtc_mask now, as shown here: https://github.com/moneroexamples/xmregcore/blob/d6601b5286ae7d98d573cc1a43dedaa3f852363e/src/UniversalIdentifier.cpp#L185


the rtc amount value is same as before. Its short now, so its padded with zero to require length.

**Comment by ndorf** [comment]
In my tests, this occurs anytime a coinbase output is used as an input to the new transaction, and not otherwise.

That includes sending any amount from a wallet containing only coinbase outs, but also from a wallet that contains a mix of smaller non-coinbase outs and larger coinbase ones: sending a larger amount (that would require using the coinbase outs as inputs) fails, but smaller amounts work fine.

All of the wallets I tested with contain only pre-fork outputs -- i.e., with a 32-byte amount and explicit 32-byte mask. So, although the format of the rct field will still be an issue for newer inputs, I don't think it is related to this error.

**Comment by moneroexamples** [comment]
@ndorf 

For me it seems to be working all fine now, and I'm pretty sure I have pre-v11 coinbase txs in my wallet. Do you have any wallet stagenet or testnet seed so that I can try to import it and check? Alternatively, you can use stagenet version of openmonero. I already updated it to v11 with the patched mymonero core js utils for testing: http://139.162.60.17:81/

If coinbase is causing problems, are you sure that this issue was solved (https://github.com/moneroexamples/openmonero/commit/fb20cb2179db6b6bda5285f432c7d10905501035/), or not reverted back in some commit? Before mymonero was also not able to spent any coinbase txs aswell, and what you describe seems as the same issue.

**Comment by moneroexamples** [comment]
Oh, now I learned that there is also some issue in monero itself with coinbase txs: https://github.com/monero-project/monero/pull/5217 

So maybe this is related and causing problems?

**Comment by luigi1111** [comment]
That should be unrelated as far as I know.

**Comment by moneroexamples** [comment]
Are you sure. Because I see that the new monero issue is that it decodeRtc for coinbase txs, when it should not https://github.com/monero-project/monero/pull/5217/commits/fa2fbc3917774ee9615e1bcf1437d57ab400018f#diff-9d580668dab930045943902048e2ac22R11322

So maybe in openmonero I dont have this issue at all, because I've been always checking if I deal with coinbase or non-coinbase txs when decoding outputs: https://github.com/moneroexamples/xmregcore/blob/master/src/UniversalIdentifier.cpp#L166

**Comment by luigi1111** [comment]
Yes. Coinbase transactions don't normally have rtc fields. Unless the failing outputs have some wacky stuff going on, this is irrelevant.

**Comment by ndorf** [comment]
> Oh, now I learned that there is also some issue in monero itself with coinbase txs: [monero-project/monero#5217](https://github.com/monero-project/monero/pull/5217)

That change fixes a potential vulnerability stemming from maliciously crafted coinbase txes. Unless I'm misunderstanding something, it doesn't have anything to do with this error (after all, even before that change, there are no such errors when sending from the official wallet).

I think you're right in that you don't have to worry about that vulnerability, if you never decode the RCT amounts of coinbase transactions to begin with.

**Comment by moneroexamples** [comment]
@ndorf 

Yesterday I forked mymonero-core-cpp, so I can push the patch: https://github.com/moneroexamples/mymonero-core-cpp/commit/993cd4f07806ed042f2b22680e80ce75cebe8116 But it has the same change as before - do not decrypt  rtc mask as encrypted one is provided by the backend. For coinbase txs, the the rtc mask is identity mask anyway, so it didn't need decrypting in the first place. Thus probably that is why coinbase txs worked for v11.

To fully check on pre-v11 coinbase txs, I will spin up new v9 stagenet node or make new private stagenet.

**Comment by ndorf** [comment]
@moneroexamples I think I have a wallet you can use, give me a few minutes.

If you do want to spin up a v9 stagenet node, you can do it much faster than normal, if you first run `monero-blockchain-export` on your existing node and then `monero-blockchain-import --dangerous-unverified-import` on the new one -- making sure to use the v0.13 version of the import utility. It will error out when it reaches the first RCTv2 tx; then you can just sync as normal (reorg should be 57 blocks or so, IIRC).

**Comment by ndorf** [comment]
@moneroexamples https://github.com/ndorf/monero_tests/blob/master/testnet-v9-coinbase.seed

**Comment by moneroexamples** [comment]
@ndorf 
Thanks. I imported the wallet, and so far don't have any issue making txs. I did not relay them (dont want to destroy your wallet), but they deamon did not complained about them and I saw them in mempool. I can relay them, just to be sure if this wallet is disposable?

I don't have testnet version of openmonero running, so for now I can paste what `get_unspend_outputs` looks like for that particular wallet which cointans `rtc` field: https://pastebin.com/YvPgf2ps  

If it was helpful, I could setup testnet version of openmonero online, so you can easly test it as well?

**Comment by ndorf** [comment]
OK, so my issue must be an unrelated regression with respect to https://github.com/mymonero/mymonero-app-js/issues/240. Sorry for the distraction, and thanks for helping!

Back to the question of these RingCT outputs, since only the 40 bytes are required for them, it would be best if we only send those. The client just needs to handle that properly (`decodeRct(..., true)`).

Using the same field and deciding based on length, as luigi [suggests](https://github.com/mymonero/mymonero-app-js/issues/277#issuecomment-468611953) seems fine, I think.

**Comment by moneroexamples** [comment]
No problem.

You mean modifying this part in mymonero-core-cpp: 

https://github.com/mymonero/mymonero-core-cpp/blob/71e383096f3c0dadebe729d61960085035f6e7b5/src/monero_transfer_utils.cpp#L164-L169

So that for v11 txs, `decodeRct` is used to decrypt the rtc mask?

But anyway, everything is fine with me, as long as the changes will be made clear so that I can adjust openmonero accordingly.

**Comment by ndorf** [comment]
I guess I meant modifying that code to use `rct::genCommitmentMask()` for v11 txes, which is what `rct::decodeRct()`, via `rct::ecdhDecode()`, would do.

That creates a conflict with our workaround for #240, though. We can't send the identity mask if we're not sending any mask at all. However, we don't actually need to send _any_ RingCT data for coinbase transactions. Therefore I propose we just use the magic string "coinbase" for those. So to summarize, we would have four possible values of the `rct` field:

1. null/undefined/empty: non-RingCT output (i.e, from version 1 tx)
2. string "coinbase" (length 8): RingCT coinbase output
3. string length 192: non-coinbase RingCT version 1 output with 256-bit amount and mask
4. string length 80: non-coinbase RingCT version 2 output with 64-bit amount and no mask

I think this covers all cases correctly with no sending of superfluous data.

For compatibility, the next client release could still support the `encrypted_mask == rct::identity()` hack, but after that it could (and I think should) be removed.

Thoughts on this? Am I missing anything?

**Comment by ndorf** [comment]
I think something like [this](https://github.com/mymonero/mymonero-core-cpp/compare/master...ndorf:feature/rct_v2) is what we would need client side -- that's not fully tested yet, though.

**Comment by moneroexamples** [comment]
Sounds good. I will try to implement these changes in openmonero and use the updated mymonero-core-cpp and see how it goes.

**Comment by moneroexamples** [comment]
@ndorf 
I implemented changes for point 4 (string length 80: non-coinbase RingCT version 2 output with 64-bit amount and no mask) for now, and testing this. In this point, frontend generated decrypted mask and am getting same error as before `tx verification failed: invalid input`. 

I think rtc_mask decoding procedure is incorrect. The `rtc_mask` value obtained from this procedure here : https://github.com/ndorf/mymonero-core-cpp/blob/feature/rct_v2/src/monero_transfer_utils.cpp#L169 is simply different from what it was before and daemon is rejecting it. 

To rectify this I modified point 4 into: string length 144: non-coinbase RingCT version 2 output: <64 rct commit as before> + <64 correctly decrypted mask by the backend> + <16 rct amount>

Decrypted mask is done by the backend anyway, so maybe better to use it, rather then re-do the same calculations on the front-end? 

I haven't checked other points yet. 

What do you think about this?

**Comment by moneroexamples** [comment]
I pushed a version with modified point 4: https://github.com/ndorf/mymonero-core-cpp/commit/84a4fe6bd3b5d7e45b406a805a078f09eac526a6 

I started testing point 2 `string "coinbase" (length 8): RingCT coinbase output`. Seems to be ok for now.

**Comment by luigi1111** [comment]
@moneroexamples yes in order to "decrypt" successfully, the front end would need the same routine added. I'm not sure which I prefer at this point (and I also need some sleep). :) I think it's probably "more correct" to have the front end regenerate the deterministic mask, though it's likely not much different in practice. Cost to regenerate is pretty trivial though.

**Comment by ndorf** [comment]
(tl;dr: one more change at https://github.com/ndorf/mymonero-core-cpp/commit/7f681b75192043b73bd23c5055011d49fbdd4240 fixes everything, I think)

@moneroexamples the elegant simplicity of your change on the OpenMonero side did not escape me. With the decrypted mask already being available in the right place at the right time, it's the best kind of 1-liner: a (-1)-liner. Unfortunately the situation with the MyMonero backend is quite the opposite. It's not infeasible, but it's a lot riskier than the solution I'm proposing here, especially given only a couple of days left to test everything. (To just scratch the surface, there is no decoding of transaction data, RingCT or otherwise, in the HTTP/API server itself.)

If it were the clearly correct solution, that might be a poor excuse, but given that it should be trivial for the client to compute the deterministic mask itself, there is no need to send that from the server for every output, I think that is preferable in and of itself.

> I think rtc_mask decoding procedure is incorrect. The rtc_mask value obtained from this procedure here : https://github.com/ndorf/mymonero-core-cpp/blob/feature/rct_v2/src/monero_transfer_utils.cpp#L169 is simply different from what it was before and daemon is rejecting it.

It's definitely incorrect, otherwise it would be working :) what you might have missed is that although it is "different from what it was before," it must be different to be correct: see [here](https://github.com/monero-project/monero/blob/5bbbe3902b4ee77ca1eb23edc0b5495812353b1f/src/ringct/rctOps.cpp#L712)

Luckily @paulshapiro was able to quickly notice that I made a mistake in my first change, and incorrectly applied an extraneous `hash_to_scalar()` to the derived key. After [fixing that](https://github.com/ndorf/mymonero-core-cpp/commit/7f681b75192043b73bd23c5055011d49fbdd4240), it looks like everything is working. I still have a lot more testing to do, but I've successfully sent a few RCT Type 4 outputs already.

**Comment by moneroexamples** [comment]
@ndorf @luigi1111 

Thank for the feedback. The new changes too `mymonero-core-cpp` seem to be working now with original point 4 (80 characters long rtc field). So now I think the main thing to do is just more testing.

**Comment by paulshapiro** [comment]
Hey @moneroexamples, for the purpose of triggering the core-cpp code to use the updated transaction construction method at the correct time, we have to add fork height info to get_unspent_outs now. We'll follow up shortly with info on how we're doing that. The parsing of these values will remain encapsulated within the core-cpp code.

**Comment by ndorf** [comment]
@moneroexamples https://github.com/monero-project/meta/pull/287#issuecomment-470172708

> Also, I'd like to propose three new fields in the `get_unspent_outs` response: `block_height`, `next_fork_block_height` and `next_fork_version`.
> 
> These would be for use by the client to determine whether or not to use, for example, RingCT type 4 when constructing new transactions.

What do you think?

**Comment by vtnerd** [comment]
> In my tests, this occurs anytime a coinbase output is used as an input to the new transaction, and not otherwise.

Crud, light-wallet-server has a similar bug where the mask field should be initialized to the identity instead of zero.

> null/undefined/empty: non-RingCT output (i.e, from version 1 tx)
> string "coinbase" (length 8): RingCT coinbase output
> string length 192: non-coinbase RingCT version 1 output with 256-bit amount and mask
> string length 80: non-coinbase RingCT version 2 output with 64-bit amount and no mask
> I think this covers all cases correctly with no sending of superfluous data.

I would consider changing the name from `rct` to distinguish the new field more clearly. Although perhaps older clients will incorrectly think its pre-ringct?

Either way, the scenario is just three: (1) no `rct` field pre-ringct, (2) `rct` field containing unencrypted mask which might be the (fixed/constant) identity mask for coinbases and (3) `"rct": "implied"` meaning the client can compute the ringct mask from the other information. If it helps the client at all, everything can be "squished" to just (1) and (2) but the server would have to compute the mask on the fly each time or store it. The server already is storing the mask _somehow_ for these older transactions, but may lose space savings on newer transactions.

I see no reason to send the amount again, its already being sent in another field. Differentiating between coinbases is possible, but leaves an extra case for the client to handle and the only penalty are the transmission bytes for sending the identity mask. The server has a fixed string constant to send either way. Unless you were proposing to send the encrypted mask, instead of unencrypted?

**Comment by moneroexamples** [comment]
@ndorf @paulshapiro 

The new field is ok. I think its main use would be to allow deployment of new client now. Since mainnet is on v9 still, backend would sent "v9" or just "9" to the client to indicate using type 3 rtc txs. When mainnet hits v10 and then v11, clients would automatically switch to type 4 rtc because backend would indicate new version of the hf.

**Comment by ndorf** [comment]
> I would consider changing the name from `rct` to distinguish the new field more clearly. Although perhaps older clients will incorrectly think its pre-ringct?

Yeah, the format is exactly the same for rct types 1-3, so old clients can continue to just work (until the actual fork, of course.) There didn't see much risk in reusing the same field for the new format as well: any reasonable client would just fail to parse it, but the worst case scenario for even an unreasonable one is an invalid tx that will be rejected.
 
> Either way, the scenario is just three: (1) no `rct` field pre-ringct, (2) `rct` field containing unencrypted mask which might be the (fixed/constant) identity mask for coinbases and (3) `"rct": "implied"` meaning the client can compute the ringct mask from the other information. If it helps the client at all, everything can be "squished" to just (1) and (2) but the server would have to compute the mask on the fly each time or store it. The server already is storing the mask _somehow_ for these older transactions, but may lose space savings on newer transactions.
> 
> I see no reason to send the amount again, its already being sent in another field. Differentiating between coinbases is possible, but leaves an extra case for the client to handle and the only penalty are the transmission bytes for sending the identity mask. The server has a fixed string constant to send either way. Unless you were proposing to send the encrypted mask, instead of unencrypted?

I'm definitely proposing that we continue sending the mask encrypted, as we are already doing. It was already [suggested](https://github.com/mymonero/mymonero-app-js/issues/277#issuecomment-468887091) that this be changed to decrypted, but that is not exactly an overnight change for the MyMonero backend. (It's also not backward compatible, for what little that's worth.)

It seems obvious now that the commitment shouldn't be needed either, but I hadn't realized it at the time. As for the encrypted amount, I'm pretty sure the MyMonero client will happily accept either format without it, so we can amend this update accordingly if you prefer. I hadn't mentioned that before now simply because I didn't want to muddy the waters. The purpose of this update wasn't to be perfect (just in case that's not obvious), but just to get the new-style outputs working in some way that everyone can implement, test and release in the couple of days left before the fork. I think it meets that goal, in spite of its rough edges. Do you disagree?

What I'm asking here is whether we can all live with the change as described here until the dust settles from this weekend's fork. At this point, @moneroexamples and I have each spent considerable time testing the updated client against our respective servers. Going back to square one is not only unappealing, to put it mildly, but given that the iOS version of the client still has to be reviewed and approved by Apple, basically means that there would be zero chance of having a working client on that platform in time.

Longer-term, there's no question that this entire field could use revisiting -- just for starters, there doesn't seem to be any good reason to be sending a blob instead of normal fields -- but what I'm suggesting (ok, insisting) is that we save that for a time when it can be discussed and done right to everyone's satisfaction without the time pressure of an imminent fork.

**Comment by ndorf** [comment]
@moneroexamples 

> The new field is ok. I think its main use would be to allow deployment of new client now. Since mainnet is on v9 still, backend would sent "v9" or just "9" to the client to indicate using type 3 rtc txs. When mainnet hits v10 and then v11, clients would automatically switch to type 4 rtc because backend would indicate new version of the hf.

Exactly right, that's its only purpose. If the field is missing, the client will just assume the latest fork rules, as it would otherwise. So, it's optional; if you don't set it the new client won't work _before_ the fork, but will afterwards, just like without the change.

**Comment by vtnerd** [comment]
> The purpose of this update wasn't to be perfect (just in case that's not obvious), but just to get the new-style outputs working in some way that everyone can implement, test and release in the couple of days left before the fork.

The format didn't actually need to be changed if client side deployment is still an issue. If you are having issues with iOS updates, I suggest looking into those changes server side, which will have some additional computation overhead until the client updates get deployed.

> I think it meets that goal, in spite of its rough edges. Do you disagree?

Yes, I disagree. We've now got multiple jacked up formats and never needed to change the format to meet the hardfork requirements. The only benefit is that server doesn't have to choose, _right now_, between a computation or storage hit to backwards support the format. So I regret for not jumping in sooner to this thread. Light-wallet-server was always storing the computed mask, so I didn't even see a reason to start a discussion about an API change when I updated it.

I'm certainly not going to recommend changes at this time - hopefully the special cases for the client can be smoothed out later.

**Comment by vtnerd** [comment]
Also, I apologize for the initial jacked up encoding scheme here. Its clearly crummy, and had it been "correct" from the beginning it would've been easier to make changes

**Comment by moneroexamples** [comment]
@vtnerd @ndorf 

I think its already too late to change the `rct` field, unless some bugs or issues are found. 

But I think the `rct` field needs to be revisited anyway as it still needs some consideration. Two days ago when I was doing tests I found that there are still problems. I think they are not critical, but the problem is that `rct` field occurs in 2 api calls: `get_unspent_outputs` and `get_random_outs`. 

So once we settled on new 4-point format of `rct` field for `get_unspent_outputs`, I naturally amended `get_random_outs` to also use the new format. The mysql-core-js was freezing at this time after the change - my guess, it does not account for format changes in  `rct` field in  `get_random_outs`. Maybe its already fixed, haven't checked yet? Or maybe my amendments in openmonero were wrong. Had not time to investigate further, so I just reverted to old format in  `get_random_outs` and it's been working fine.

So for now it seems that `rct` has to have two different formats. New format for `get_unspent_outputs` and old one for `get_random_outs`. I wonder how `get_random_outs` is handled in MyMonero? Does it still uses old format for `rct` field?

p.s.

What will be the new format for the `fork_version` field?

**Comment by vtnerd** [comment]
Another goof in the existing format, only the commitment is needed for the `get_random_outs` endpoint.

The new format cannot be used for `get_random_outs` if a coinbase is selected as a dummy output - which is very likely for testnet. The JS should get some exception deep within a cpp portion (or some such thing). The JS should be unable to compute a valid CT proof since it doesn't know the "dummy output" commitment. I'd have to look at the equations again, maybe the signer doesn't need the other commitments, but that doesn't seem correct.
 
Use the existing 96-character format, and put _any_ valid hex values for the amount and mask portion. The client is unable to decrypt them anyway, and they aren't needed for the signature. Currently the JS should be ignoring everything but the first 64-hex characters. The light-wallet-server is actually sending an invalid format containing just the first 64 hex characters, and (I'm fairly certain that) the JS client code accepted this at some point. It looks like `wallet2` requires the first 128-hex characters, crud.

**Comment by vtnerd** [comment]
Also the new field _could_ work if `get_random_outs` returned an amount field for each output or more specifically coinbase outputs. The amount is stored as `amount == 0` in the DB, which is why they get returned on calls to 0, but the amount is publicly known and not zero. The client could then compute the "zero-commitment" based on the amount specifically for coinbase outputs.

**Comment by ndorf** [comment]
@moneroexamples `fork_version` is an integer in the range 1-255. It should be 9 now, and 10 after block 1788000 (on mainnet). 

I realize now that the API document is not clear on the format of the `get_random_outs` response. The MyMonero server also sends only the commitment (32 bytes, or 64 hex chars) in those, no amount or mask.

Very concerning though, that you got the client to hang. I'll try to reproduce that in my tests.

**Comment by ndorf** [comment]
@vtnerd I see your point now. Yes, we could have made it work without any client-side change by encrypting the generated deterministic mask and sending that. Perhaps that would have been preferable after all.

I take it that's what light_wallet_server is doing already? (The branch wasn't up to date yet, when I tried to look.) If so, that will continue to work, since that format remains exactly the same for rct v1.

**Comment by paulshapiro** [comment]
@moneroexamples core-js and core-cpp have been updated with fork_version support

**Comment by moneroexamples** [comment]
@ndorf 

Probably `rct` field in `get_random_outs` should be renamed to "commitment" or something similar. As @vtnerd pointed out the problem I experience with `get_random_outs` and mymonero-core-js was that I assumed that `rct` is same as in `get_unspent_outputs`, and it was hanging at "coinbase" string probably. I didn't investigated this issue much at the time. Once there will be some peaceful time after the hf, maybe can revisit this issue.

@paulshapiro 
Thanks for updated. I will pull the new changes and check how they work with new `fork_version` field.

**Comment by vtnerd** [comment]
> I take it that's what light_wallet_server is doing already? (The branch wasn't up to date yet, when I tried to look.) If so, that will continue to work, since that format remains exactly the same for rct v1.

Yes, light-wallet-server is storing the mask unencrypted in its DB then computes the `rct` field on-the-fly. The mask is still computed for `amount` decryption anyway, so its still stored. The one bug is that coinbases are set to `rct::zero()` instead of `rct::identity()`, which will get pushed out next.
